### DISCUSSION TITLE
빌드 스크립트

### DISCUSSION THREAD

**Discussion by x-0o0** [ku-ring/app-ios-v2#56] [CI/CD]
## 개요

빌드가 되는 유효한 코드가 머지될 수 있도록 다음의 빌드 스크립트가 필요함
- KuringApp
- KuringModulePackage 의 각 라이브러리

### 체크항목

- [x] 패키지 빌드 성공여부 체크
- [x] iOS 최소 버전에서 앱 타겟 빌드 성공 여부 체크
- [x] iOS 최신 버전에서 앱 타겟 빌드 성공 여부 체크

- [x] 앱 테스트 성공 여부 체크
- [x] 패키지 테스트 성공 여부 체크

### 효율
- [ ] KuringModulePackage 의 모든 라이브러리를 빌드하는 것이 아닌
수정이 있는 라이브러리만 빌드할 수 있도록
- [x] 불필요하게 깃헙 액션 사용시간이 소모되는 것을 방지하기 위해 커멘트로 트리거 하도록 수정
- [x] 2approval 없어도 머지하는 테스트 추가 (간단한 변경사항의 경우 불필요한 시간닝비를 막기위함)

### 트리거 명령어
- /쿠링 ios17 앱 빌드
- /쿠링 패키지 빌드
- /쿠링 패키지 테스트
- /~~테스트~~
- /~~전부 체크~~
- /쿠링 머지

**추가 사용법**
- /쿠링 패키지 빌드하고 머지
- /쿠링 ios17 앱 빌드하고 머지

**Comment by x-0o0** [comment]
커멘트로 깃헙 액션 작동시키는 스크립트 작성은 

[cozzin](https://github.com/cozzin) 님이 작성한 블로그 포스트

https://medium.com/@hongseongho/github-action으로-comment-bot-만들기-422e6e471c8e

에 잘 정리되어있음

**Comment by x-0o0** [comment]
## 마주한 문제:

### 문제 요약
분명 스크립트는 잘 작성했는데 커멘트로 트리거가 되지 않는 현상이 있었다.

### 원인
> "Note: This event will only trigger a workflow run if the workflow file is on the default branch."
> [출처 | docs.github.com](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#public)

`issue-comment` 트리거는 **default branch** 기반으로 돌아가기 때문에 아직 `yaml` 파일이 **default branch** 에 머지되지 않은 시점에서 `issue-comment` 트리거 시도를 해도 **default branch 에는 트리거할 액션이 정의되어 있지 않아서 동작하지 않았다.**

### 해결방안
테스트 할때는 PR 브랜치로 default branch를 임시로 변경하고, 확인이 끝나면 팀원들에게 공유 후 default branch 복구

**Comment by x-0o0** [comment]
## 마주한 문제

### 문제 요약

스크립트에 테스트가 최신 Xcode 버전에서 돌아가도록 작성했는데, Xcode15 가 아닌, Xcode14.2 (심지어 14.3 도 아님) 으로 돌아가는 문제가 있었다.

### 원인
`macos-latest` 를 사용하는 경우 맥OS 에 Applications 폴더에는 Xcode_14.2.app 까지만 설치 되어있다.

### 해결 방안

> Looks like you need to use macos-13, not macos-latest.
> [link](https://github.com/actions/runner-images/issues/7672#issuecomment-1597219552)

```
build:
-  runs-on: macOS-latest
+  runs-on: macos-15
    steps:
      - name: Setup Xcode version
        uses: maxim-lobanov/setup-xcode@v1
        with:
          xcode-version: '15.0'
```

**Comment by x-0o0** [comment]
## 마주한 문제

> **정보** 이 문제는 known-issue 테크뎁으로 둡니다.

### 문제 요약

Tests must be run on a concrete device

### 원인
`#Preview` 매크로 빌드를 위해 Xcode15 환경이 필요. 하지만 시뮬레이터 버전이 iOS 17 만 있음. (`xcrun xctrace list devices`)
시뮬을 설치하는 방법이 있으나 소모시간 너무 길고, 특히 macOS 는 깃헙 액션 사용시간이 몇배수로 적용되기 때문에 (3이었나 10이었나...)
최선의 방법은 아닌 것 같음

**Comment by x-0o0** [comment]
## 마주한 문제

### 문제 요약

깃헙 액션 테스트가 PR 커밋에서가 아닌 Base 브랜치 소스로 돌아가서 실제 PR 의 코드가 테스트 되지 않고 있음 https://github.com/ku-ring/ios-app/issues/84

### 원인
`issue_comment` 이벤트에 의해 동작하는 워크플로우는 base branch 로 체크아웃 함

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#issue_comment

![issue_comment](https://github.com/ku-ring/ios-app/assets/53814741/bb09bc58-7bda-47ce-a19a-2a87da604533)

보면 GITHUB_REF 가 default branch 입니다. 
pull_request 브랜치에서 돌리기 위해 방법 2개를 시도했으며 결과를 같이 공유합니다.

### 방법1: `pull_request_review_comment` 이벤트 사용

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_review_comment

![pull_request_review_comment](https://github.com/ku-ring/ios-app/assets/53814741/62efe431-720b-4bb8-8ef5-a0c9273725e0)

```yml
on:
  pull_request_review_comment:
    types: [created, edited]
```
로 수정하고
```yml
- if: github.event.issue.pull_request && contains(github.event.comment.body, '/쿠링') && contains(github.event.comment.body, '패키지 테스트')
+ if: contains(github.event.comment.body, '/쿠링') && contains(github.event.comment.body, '패키지 테스트') 
```

조건문에서 `github.event.issue.pull_request` 조건을 제거 합니다. 나머지 조건의 경우 event 하위 프로퍼티 동일하기 때문에 변경할 필요 없습니다.

PR reviewer 가 리뷰 커멘트를 남겼을 경우에만 동작하게 됩니다.

`pull_request_review_comment` 은 PR을 생성한 사람이 워크플로우를 트리거할 수 없다는 단점이 있습니다.

관련 작업 커밋 링크: https://github.com/ku-ring/ios-app/pull/87/commits/ddf207025e17c5e4084a741395a8633eda4b74f5

### 방법2: `issue_comment` 이벤트 시, 체크아웃에 `ref` 값으로 `event.issue.number` 사용

`issue_comment` 로 해결하기 위해 `issue_comment` 이벤트 시, PR 브랜치 최신 커밋(`HEAD`) 로 체크아웃 하는 방법에 대해서 리서치 하던 중 다음 답변을 발견했습니다.

https://github.com/actions/checkout/issues/331#issuecomment-1438220926

```yml
- name: Checkout pull request 🏁
        uses: actions/checkout@v3 # v4로 적용
        with:
          ref: refs/pull/${{ github.event.issue.number }}/head
```

이를 base 브랜치에서 반영 후 PR #90 에서 테스트를 실행시켜서 다음과 같이 PR 브랜치로 잘 체크아웃 되는 것을 확인하였습니다.

![method2_result](https://github.com/ku-ring/ios-app/assets/53814741/a720bddf-2335-4bfe-ae93-1e728ac5f5a8)

**Reply by sunshiningsoo** [reply]
대박입니다,,, 배워가욥!!!

**Reply by lgvv** [reply]
👀 CI/CD 공부하는데 진짜 도움 많이받고 있습니다!!
